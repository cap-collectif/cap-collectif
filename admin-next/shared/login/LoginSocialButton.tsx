import React, { useState } from 'react'
import { FormattedMessage, useIntl } from 'react-intl'
import styled from 'styled-components'
import tinycolor from 'tinycolor2'
import { Box, Link, Text } from '@cap-collectif/ui'
import SocialIcon from '@shared/ui/LegacyIcons/SocialIcon'
import getBaseUrl from '@shared/utils/getBaseUrl'

export type LoginSocialButtonType = 'facebook' | 'openId' | 'franceConnect' | 'saml' | 'cas' | 'oauth2'
type Props = {
  type: LoginSocialButtonType
  switchUserMode?: boolean
  text?: string
  justifyContent?: string
  noHR?: boolean
  primaryColor?: string
  colorText?: string
  fcTitle?: string
}
export const getLabelColorForType = (type: LoginSocialButtonType, color?: string): string => {
  switch (type) {
    case 'openId':
      return color || 'white'

    case 'facebook':
    case 'saml':
    case 'cas':
    case 'franceConnect':
      return 'white'

    default:
      return 'white'
  }
}
export const getButtonColorForType = (type: LoginSocialButtonType, bgColor?: string): string => {
  switch (type) {
    case 'facebook':
      return '#3B5998'

    case 'openId':
      return bgColor || '#1b9bd1'

    case 'saml':
      return '#7498c0'

    case 'cas':
      return '#7498c0'

    case 'franceConnect':
      return '#034ea2'

    default:
      return '#034ea2'
  }
}
export const getButtonLinkForType = (
  type: LoginSocialButtonType,
  redirectUri: string,
  isInvitationSSO: boolean = false,
): string => {
  const destination = isInvitationSSO ? getBaseUrl() : window && window.location.href

  switch (type) {
    case 'facebook':
      return `/login/facebook?_destination=${destination}`

    case 'openId':
    case 'oauth2':
      return `/login/openid?_destination=${redirectUri}`

    case 'cas':
      return `/login-cas?_destination=${destination}`

    case 'saml':
      return `/login-saml?_destination=${destination}`

    case 'franceConnect':
      return redirectUri
        ? `/login/franceconnect?_destination=${redirectUri}`
        : `/login/franceconnect?_destination=${destination}`

    default:
      return ''
  }
}
export const getButtonContentForType = (type: string): string => {
  switch (type) {
    case 'facebook':
      return 'Facebook'

    case 'saml':
      return 'Saml'

    case 'cas':
      return 'ARTUR'

    default:
      return ''
  }
}
export type LinkButtonProps = {
  type: LoginSocialButtonType
  labelColor?: string
  buttonColor?: string
  maxWidth?: string
  text?: string
  children: any
}
const LinkButton = styled.div<LinkButtonProps>`
  position: relative;
  margin-top: 10px;
  height: 34px;
  width: 100%;
  max-width: ${props => `${props.maxWidth}`};
  border-radius: 3px;
  display: flex;

  && {
    color: ${props => getLabelColorForType(props.type, props.labelColor)};
    background-color: ${props => getButtonColorForType(props.type, props.buttonColor)};
  }

  .loginIcon {
    top: 0;
    fill: ${props => getLabelColorForType(props.type, props.labelColor)};
    background-color: ${props => tinycolor(getButtonColorForType(props.type, props.buttonColor)).darken(10).toString()};
    height: 34px;
    width: 15%;
    border-radius: 3px 0 0 3px;
    display: flex;
    align-items: center;
    justify-content: center;

    & > svg {
      height: 60%;
      width: 100%;
      ${props => {
        if (props.type === 'openId' || props.type === 'saml' || props.type === 'cas') {
          return 'transform: translate(3px, 2px) scale(1.3);'
        }
      }}
    }
  }

  a {
    width: 100%;
    text-decoration: none;
    color: ${props => getLabelColorForType(props.type, props.labelColor)};
    display: flex;
    align-items: center;

    &:focus,
    &:hover {
      color: ${props => getLabelColorForType(props.type, props.labelColor)};
    }

    span {
      margin-left: 16px;
    }
  }

  &:focus,
  &:hover {
    background-color: ${props => tinycolor(getButtonColorForType(props.type, props.buttonColor)).darken(10)};
    .loginIcon {
      background-color: ${props => tinycolor(getButtonColorForType(props.type, props.buttonColor)).darken(20)};
    }
  }
`

export const FranceConnectButton = styled.div<{
  justifyContent: string
}>`
  position: relative;
  margin-top: 10px;
  width: 100%;
  display: flex;
  justify-content: ${props => props.justifyContent};
  a[title] {
    text-transform: uppercase;
  }
  .loginIcon {
    & > svg {
      height: 60px;
      width: 228px;
    }
  }
`
const GrandLyonConnectButton = styled.div`
  position: relative;
  margin-top: 10px;
  width: 100%;
  display: flex;
  justify-content: center;
  a {
    width: 100%;
  }
`

const FC_COLOR = '#000091'

export const LoginSocialButton = ({
  type,
  switchUserMode,
  text,
  justifyContent = 'center',
  noHR = false,
  primaryColor,
  colorText,
  fcTitle = 'fc-title',
}: Props) => {
  const [isHover, seIsHover] = useState<boolean>(false)
  const redirectUri = switchUserMode ? `${getBaseUrl()}/sso/switch-user` : window && window.location.href
  const intl = useIntl()

  if (text === 'grandLyonConnect') {
    return (
      <GrandLyonConnectButton>
        <a href={getButtonLinkForType(type, redirectUri)} title={type}>
          <SocialIcon className="loginIcon" name={text} />
        </a>
      </GrandLyonConnectButton>
    )
  }

  const isFcRequirement = fcTitle === 'fc-requirement-title'
  return (
    <div>
      {type === 'franceConnect' ? (
        <>
          <Text
            // @ts-ignore rapatrier DS et fix
            textAlign={justifyContent}
            mt={1}
            lineHeight="s"
            fontSize={isFcRequirement ? '14px' : 2}
            color={isFcRequirement ? 'gray.700' : undefined}
          >
            <FormattedMessage tagName="p" id={fcTitle} />
          </Text>
          <FranceConnectButton justifyContent={justifyContent}>
            <a
              href={getButtonLinkForType(type, redirectUri)}
              title="FranceConnect"
              onMouseEnter={() => seIsHover(true)}
              onMouseLeave={() => seIsHover(false)}
            >
              {isHover ? (
                <SocialIcon className="loginIcon" name={`${type}Hover`} />
              ) : (
                <SocialIcon className="loginIcon" name={type} />
              )}
            </a>
          </FranceConnectButton>
          <Box mt={2} fontSize={2}>
            <Link
              color={FC_COLOR}
              _hover={{ color: FC_COLOR }}
              href="https://franceconnect.gouv.fr/"
              target="_blank"
              rel="noreferrer"
            >
              <FormattedMessage id="what-is-fc" />
            </Link>
            {!noHR ? (
              <Text textAlign="center" as="div" fontWeight="bold" color="neutral-gray.700" my={5}>
                {intl.formatMessage({
                  id: 'login.or',
                })}
              </Text>
            ) : null}
          </Box>
        </>
      ) : (
        <LinkButton type={type} labelColor={colorText} buttonColor={primaryColor}>
          <SocialIcon className="loginIcon" name={type} />
          <a href={getButtonLinkForType(type, redirectUri)} title={type}>
            {text !== undefined ? <span>{text}</span> : <FormattedMessage id={getButtonContentForType(type)} />}
          </a>
        </LinkButton>
      )}
    </div>
  )
}

export default LoginSocialButton
