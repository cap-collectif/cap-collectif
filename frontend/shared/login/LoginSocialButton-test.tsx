/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { LoginSocialButton } from './LoginSocialButton'

const baseProps = {
  primaryColor: '#fffeee',
  colorText: '#fff',
}
describe('<LoginSocialButton />', () => {
  it('should render correctly', () => {
    const props = { ...baseProps, type: 'facebook', switchUserMode: false }
    const wrapper = shallow(<LoginSocialButton {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should render correctly franceConnect', () => {
    const props = { ...baseProps, type: 'franceConnect', switchUserMode: false }
    const wrapper = shallow(<LoginSocialButton {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should render correctly with grandLyonConnect', () => {
    const props = { ...baseProps, type: 'openId', text: 'grandLyonConnect' }
    const wrapper = shallow(<LoginSocialButton {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should render correctly with switchusermode', () => {
    const props = { ...baseProps, type: 'openId', switchUserMode: true, text: 'openid' }
    const wrapper = shallow(<LoginSocialButton {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
