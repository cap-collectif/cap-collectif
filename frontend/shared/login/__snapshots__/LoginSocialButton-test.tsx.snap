// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginSocialButton /> should render correctly 1`] = `
<div>
  <LoginSocialButton__LinkButton
    buttonColor="#fffeee"
    labelColor="#fff"
    type="facebook"
  >
    <SocialIcon
      className="loginIcon"
      name="facebook"
    />
    <a
      href="/login/facebook?_destination=http://localhost/"
      title="facebook"
    >
      <FormattedMessage
        id="Facebook"
        values={Object {}}
      />
    </a>
  </LoginSocialButton__LinkButton>
</div>
`;

exports[`<LoginSocialButton /> should render correctly franceConnect 1`] = `
<div>
  <Text
    fontSize={2}
    lineHeight="s"
    mt={1}
    textAlign="center"
  >
    <FormattedMessage
      id="fc-title"
      tagName="p"
      values={Object {}}
    />
  </Text>
  <LoginSocialButton__FranceConnectButton
    justifyContent="center"
  >
    <a
      href="/login/franceconnect?_destination=http://localhost/"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      title="FranceConnect"
    >
      <SocialIcon
        className="loginIcon"
        name="franceConnect"
      />
    </a>
  </LoginSocialButton__FranceConnectButton>
  <LoginSocialButton__FranceConnectLink
    justifyContent="center"
  >
    <a
      href="https://franceconnect.gouv.fr/"
      rel="noreferrer"
      target="_blank"
    >
      <FormattedMessage
        id="what-is-fc"
        values={Object {}}
      />
    </a>
    <Text
      as="div"
      color="neutral-gray.700"
      fontWeight="bold"
      my={5}
      textAlign="center"
    >
      login.or
    </Text>
  </LoginSocialButton__FranceConnectLink>
</div>
`;

exports[`<LoginSocialButton /> should render correctly with grandLyonConnect 1`] = `
<LoginSocialButton__GrandLyonConnectButton
  type="grandLyonConnect"
>
  <a
    href="/login/openid?_destination=http://localhost/"
    title="openId"
  >
    <SocialIcon
      className="loginIcon"
      name="grandLyonConnect"
    />
  </a>
</LoginSocialButton__GrandLyonConnectButton>
`;

exports[`<LoginSocialButton /> should render correctly with switchusermode 1`] = `
<div>
  <LoginSocialButton__LinkButton
    buttonColor="#fffeee"
    labelColor="#fff"
    type="openId"
  >
    <SocialIcon
      className="loginIcon"
      name="openId"
    />
    <a
      href="/login/openid?_destination=http://localhost/sso/switch-user"
      title="openId"
    >
      <span>
        openid
      </span>
    </a>
  </LoginSocialButton__LinkButton>
</div>
`;
